src/sdk/pynni/nni/smac_tuner/smac_tuner.py

```diff
@@ -226,8 +228,16 @@ def generate_parameters(self, parameter_id, **kwargs):
         else:
             challengers = self.smbo_solver.nni_smac_request_challengers()
             for challenger in challengers:
+                if self.dedup:
+                    match = [v for k, v in self.total_data.items() \
+                             if v.get_dictionary() == challenger.get_dictionary()]
+                    if match:
+                        continue
                 self.total_data[parameter_id] = challenger
                 return self.param_postprocess(challenger.get_dictionary())
+            assert self.dedup is True
```

Comment:
In theory, if challengers is empty,  this assertion could fails if self.dedup is false, but in this situation, we want to raise NoMoreTrialError?

Reply:
according to my understanding, `challengers` would not be empty. if it turns out `challengers` is empty, I want the experiment to crash.